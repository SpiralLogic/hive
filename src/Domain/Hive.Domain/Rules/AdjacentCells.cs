using System.Collections.Generic;
using Hive.Domain.Entities;

namespace Hive.Domain.Rules
{
    public class AdjacentCells : IRule
    {
        public ISet<Coords> ApplyRule(Cell originCell, ISet<Cell> allCells)
        {
            return originCell.Coords.GetNeighbors();
        }
    }
}
