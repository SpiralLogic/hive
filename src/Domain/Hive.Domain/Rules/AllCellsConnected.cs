using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Hive.Domain.Entities;
using Hive.Domain.Extensions;

namespace Hive.Domain.Rules
{
    public class AllCellsConnected : IRule
    {
        private HashSet<Cell> checkd = new HashSet<Cell>();

        public ISet<Coords> ApplyRule(Cell currentCell, ISet<Cell> allCells)
        {
            checkd.Clear();
            return allCells.WhereEmpty().ToCoords();
        }
    }
}
