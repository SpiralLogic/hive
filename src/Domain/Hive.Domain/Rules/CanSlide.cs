using System.Collections.Generic;
using System.Linq;
using Hive.Domain.Entities;
using Hive.Domain.Extensions;

namespace Hive.Domain.Rules
{
    public class CanSlide : IRule
    {
        private readonly ISet<Cell> _checked = new HashSet<Cell>();

        public ISet<Coords> ApplyRule(Cell originCell, ISet<Cell> allCells)
        {
            _checked.Clear();
            return GetSlidableNeighbors(originCell, allCells).ToCoords();
        }

        private IEnumerable<Cell> GetSlidableNeighbors(Cell start, ISet<Cell> allCells)
        {
            _checked.Add(start);
            var neighbors = allCells.SelectNeighbors(start).ToHashSet();
            var newSlideTo = neighbors.WhereEmpty().Where(end => CanSlideTo(end, neighbors, allCells)).ToList();

            return newSlideTo.Union(newSlideTo.Except(_checked).SelectMany(c => GetSlidableNeighbors(c, allCells)));
        }

        private static bool CanSlideTo(Cell end, IEnumerable<Cell> neighbors, ISet<Cell> allCells) =>
            allCells.SelectNeighbors(end).Intersect(neighbors).WhereOccupied().Count() != 2;
    }
}
