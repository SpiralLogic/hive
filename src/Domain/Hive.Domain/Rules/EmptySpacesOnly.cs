using System.Collections.Generic;
using Hive.Domain.Entities;
using Hive.Domain.Extensions;

namespace Hive.Domain.Rules
{
    public class EmptySpacesOnly : IRule
    {
        public ISet<Coords> ApplyRule(Cell currentCell, ISet<Cell> cells)
        {
            return cells.WhereEmpty().ToCoords();
        }
    }
}
