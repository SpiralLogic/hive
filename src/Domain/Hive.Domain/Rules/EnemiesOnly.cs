using System.Collections.Generic;
using Hive.Domain.Entities;
using Hive.Domain.Extensions;
using System.Linq;

namespace Hive.Domain.Rules
{
    public class EnemiesOnly : IRule
    {
        public ISet<Coords> ApplyRule(Cell originCell, ISet<Cell> allCells)
        {
            return allCells.Where(c=> !c.IsEmpty() && c.TopTile().PlayerId != originCell.TopTile().PlayerId).ToCoords();
        }
    }
}
