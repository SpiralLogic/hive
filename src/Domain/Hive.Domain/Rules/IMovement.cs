using System.Collections.Generic;

namespace Hive.Domain.Entities
{
    internal interface IMovement
    {
        ISet<Coords> GetAvailableMoves(Coords tileCoords, ISet<Cell> cells);
    }
}