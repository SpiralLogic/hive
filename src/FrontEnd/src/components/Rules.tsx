import { FunctionComponent, h } from 'preact';
import { RuleCell } from './RuleCell';
import Cell from './Cell';
import Hextille from './Hextille';
import Row from './Row';
import Tile from './Tile';

export const BeetleRule: FunctionComponent = () => {
  return (
    <Hextille class="rules">
      <Row>
        <Cell hidden={true} />
        <RuleCell result="correct" />
        <Cell hidden={true} />
        <Cell hidden={true} />
      </Row>
      <Row class="start">
        <Cell hidden={true} />
        <RuleCell result="correct" creature="spider" />
        <RuleCell
          creature="beetle"
          class="selected"
          correct={['topLeft', 'left', 'bottomLeft', 'bottomRight']}
        />
      </Row>
      <Row>
        <Cell hidden={true} />
        <RuleCell result="correct" creature="ant" />
        <RuleCell result="correct" />

        <Cell hidden={true} />
      </Row>
      <Row>
        <Cell hidden={true} />
        <Cell hidden={true} />
        <RuleCell creature="spider" />
        <Cell>
          <Tile creature="grasshopper" />
        </Cell>
      </Row>
    </Hextille>
  );
};

BeetleRule.displayName = 'BeetleRule';

export const QueenRule: FunctionComponent = () => {
  return (
    <Hextille class="rules">
      <Row>
        <RuleCell result="correct" />
        <RuleCell creature="ant" />
        <RuleCell creature="spider" />
        <Cell hidden={true} />
      </Row>
      <Row class="start">
        <RuleCell result="correct" />
        <RuleCell
          zIndex={1}
          creature="queen"
          class="selected"
          correct={['topLeft', 'left', 'right', 'bottomRight']}
        />
        <RuleCell result="correct" />
        <RuleCell creature="spider" />
      </Row>
      <Row>
        <RuleCell creature="grasshopper" />
        <RuleCell result="correct" />
        <RuleCell creature="queen" />
        <RuleCell creature="beetle" />
      </Row>
      <Row>
        <Cell hidden={true} />
        <RuleCell creature="grasshopper" />
        <RuleCell creature="ant" />
        <RuleCell creature="beetle" />
      </Row>
    </Hextille>
  );
};

QueenRule.displayName = 'QueenRule';

export const GrasshopperRule: FunctionComponent = () => {
  return (
    <Hextille class="rules">
      <Row zIndex={1}>
        <Cell hidden={true} />
        <RuleCell result="correct" />
        <RuleCell creature="beetle" />
        <RuleCell
          creature="grasshopper"
          class="selected"
          correct={[
            ['left', 3],
            ['right', 3],
            ['bottomLeft', 5],
          ]}
          incorrect={[['bottomRight', 5]]}
        />
        <RuleCell creature="queen" zIndex={-1} />
        <RuleCell result="correct" />
      </Row>
      <Row>
        <Cell hidden={true} />
        <Cell hidden={true} />
        <RuleCell creature="grasshopper" />
        <Cell hidden={true} />
        <RuleCell creature="ant" />
      </Row>
      <Row>
        <Cell hidden={true} />
        <Cell hidden={true} />
        <RuleCell creature="queen" />
        <RuleCell creature="spider" />
        <RuleCell creature="ant" />
        <Cell hidden={true} />
      </Row>
      <Row>
        <RuleCell creature="spider" />
        <RuleCell result="correct" />
        <RuleCell creature="beetle" />
        <Cell hidden={true} />
        <RuleCell result="incorrect" />
      </Row>
      <Row>
        <Cell hidden={true} />
        <RuleCell creature="spider" />
        <RuleCell creature="spider" />
      </Row>
    </Hextille>
  );
};
GrasshopperRule.displayName = 'GrasshopperRule';

export const SpiderRule: FunctionComponent = () => {
  return (
    <Hextille class="rules">
      <Row>
        <RuleCell result="correct" />
        <RuleCell result="correct" symbol="2" correct={['left']} />
        <RuleCell class="selected" creature="spider" zIndex={5} correct={['right', 'bottomLeft']} />
        <RuleCell result="correct" symbol="1" zIndex={4} correct={['bottomRight']} />
      </Row>
      <Row zIndex={3}>
        <RuleCell creature="beetle" />
        <RuleCell result="correct" symbol="1" zIndex={3} correct={['topLeft', 'bottomLeft', 'bottomRight']} />
        <RuleCell creature="spider" />
        <RuleCell result="correct" symbol="2" zIndex={2} correct={['bottomRight']} />
      </Row>
      <Row zIndex={2}>
        <RuleCell creature="ant" />
        <RuleCell result="correct" symbol="2" zIndex={2} correct={['bottomRight']} />
        <RuleCell result="correct" symbol="2" zIndex={2} correct={['bottomLeft', 'bottomRight']} />
        <RuleCell creature="queen" />
        <RuleCell result="correct" zIndex={1} />
      </Row>
      <Row zIndex={1}>
        <RuleCell creature="queen" />
        <RuleCell result="correct" />
        <RuleCell result="correct" />
        <RuleCell creature="beetle" />
        <Cell hidden={true} />
      </Row>
      <Row>
        <RuleCell creature="grasshopper" />
        <RuleCell creature="grasshopper" />
        <RuleCell creature="ant" />
      </Row>
    </Hextille>
  );
};
SpiderRule.displayName = 'GrasshopperRule';
export const AntRule: FunctionComponent = () => {
  return (
    <Hextille class="rules">
      <Row>
        <Cell hidden={true} />
        <RuleCell result="correct" correct={['right']} zIndex={1} />
        <RuleCell result="correct" />
        <Cell hidden={true} />
        <Cell hidden={true} />
      </Row>
      <Row>
        <RuleCell result="correct" correct={['topRight']} />
        <RuleCell creature="queen" />
        <RuleCell result="correct" />
        <RuleCell result="correct" correct={['left']} />
      </Row>
      <Row>
        <RuleCell result="correct" correct={['topRight']} />
        <RuleCell creature="grasshopper" />
        <Cell hidden={true} />
        <RuleCell creature="queen" />
        <RuleCell result="correct" correct={['topLeft']} />
      </Row>
      <Row>
        <RuleCell result="correct" correct={['topLeft']} />
        <RuleCell creature="beetle" />
        <RuleCell creature="beetle" />
        <RuleCell result="correct" correct={['topRight']} />
      </Row>
      <Row>
        <Cell hidden={true} />
        <RuleCell result="correct" correct={['topLeft']} />
        <RuleCell class="selected" creature="ant" correct={['left', 'right']} zIndex={1} />
        <RuleCell result="correct" correct={['topRight']} />
      </Row>
    </Hextille>
  );
};
AntRule.displayName = 'AntRule';
