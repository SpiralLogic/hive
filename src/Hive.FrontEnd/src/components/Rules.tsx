import '../css/rules.css';

import { useReducer } from 'preact/hooks';

import AntRules from './rules/AntRules';
import BeetleRules from './rules/BeetleRules';
import FreedomToMove from './rules/FreedomToMoveRule';
import GrasshopperRules from './rules/GrasshopperRules';
import Objective from './rules/Objective';
import OneHiveRule from './rules/OneHiveRule';
import QueenRules from './rules/QueenRules';
import SpiderRules from './rules/SpiderRules';

const ruleList = [
  Objective,
  QueenRules,
  BeetleRules,
  SpiderRules,
  GrasshopperRules,
  AntRules,
  OneHiveRule,
  FreedomToMove,
];
const changeRule = (currentRuleIndex: number, { type }: { type: 'next' | 'prev' }): number => {
  if (type === 'next') return (currentRuleIndex + 1) % ruleList.length;
  if (currentRuleIndex > 0) return (currentRuleIndex - 1) % ruleList.length;
  return ruleList.length - 1;
};

const Rules = () => {
  const [currentRuleIndex, changeCurrentRule] = useReducer(changeRule, 0);
  const RuleComponent = ruleList[currentRuleIndex];
  return (
    <div>
      <div class="menu">
        <button title="Previous" onClick={() => changeCurrentRule({ type: 'prev' })}>
          {`<`}
        </button>
        <h3>{RuleComponent.title}</h3>
        <button autofocus={true} title="Next" onClick={() => changeCurrentRule({ type: 'next' })}>
          {`>`}
        </button>
      </div>
      <ul>
        {RuleComponent.description.map((d) => (
          <li key={d}>{d}</li>
        ))}
      </ul>
      <RuleComponent key={RuleComponent.title} />
    </div>
  );
};

Rules.displayName = 'Rules';
export default Rules;
