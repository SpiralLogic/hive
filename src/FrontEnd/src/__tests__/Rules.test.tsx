import { h } from 'preact';
import { render, screen } from '@testing-library/preact';
import { renderElement } from './test-helpers';
import Rules from '../components/Rules';
import userEvent from '@testing-library/user-event';

describe('Rules tests', () => {
  const renderRules = () => {
    const close = jest.fn();
    const rules = render(<Rules setShowRules={close} />);
    return [rules, close];
  };

  test('next button moves next', () => {
    renderRules();
    userEvent.click(screen.getByTitle('Next'));

    expect(document.querySelector('.selected.queen')).toBeInTheDocument();
  });

  test('prev button goes back to end', () => {
    renderRules();
    userEvent.click(screen.getByTitle('Previous'));

    expect(document.querySelector('.ant')).toBeInTheDocument();
  });

  test('prev button moves back', () => {
    renderRules();
    userEvent.click(screen.getByTitle('Next'));
    userEvent.click(screen.getByTitle('Next'));
    userEvent.click(screen.getByTitle('Previous'));

    expect(document.querySelector('.selected.queen')).toBeInTheDocument();
  });

  test('snapshot', () => {
    expect(renderElement(<Rules setShowRules={() => ({})} />)).toMatchSnapshot();
  });
});
