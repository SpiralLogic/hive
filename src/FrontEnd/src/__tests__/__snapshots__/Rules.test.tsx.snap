// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rules tests snapshot 1`] = `
<div
  class="modal"
>
  <div
    class="rules"
    role="dialog"
    title="rules"
  >
    <div
      class="menu"
    >
      <button
        title="Previous"
      >
        &lt;
      </button>
      <button
        title="Next"
      >
        &gt;
      </button>
    </div>
    <div
      class="right"
      role="grid"
    >
      <div
        role="row"
      >
        <div
          role="none"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
        </div>
        <div
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile ant"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature"
                href="#ant"
              />
            </svg>
            
          </div>
        </div>
        <div
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile grasshopper"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature"
                href="#grasshopper"
              />
            </svg>
            
          </div>
        </div>
      </div>
      <div
        role="row"
      >
        <div
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile spider"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature"
                href="#spider"
              />
            </svg>
            
          </div>
        </div>
        <div
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile queen"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature"
                href="#queen"
              />
            </svg>
            
          </div>
        </div>
        <div
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile beetle"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature"
                href="#beetle"
              />
            </svg>
            
          </div>
        </div>
      </div>
      <div
        role="row"
      >
        <div
          role="none"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
        </div>
        <div
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile ant"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature"
                href="#ant"
              />
            </svg>
            
          </div>
        </div>
        <div
          role="cell"
        >
          <svg
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              href="#hex"
            />
          </svg>
          <div
            class="tile grasshopper"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature"
                href="#grasshopper"
              />
            </svg>
            
          </div>
        </div>
      </div>
    </div>
    <caption>
      The first player to have a surrounded queen loses.
      <br />
       New creatures entering the game can only touch friendly creatures.
    </caption>
  </div>
</div>
`;
